pyboy/core/mb.py

```diff
@@ -122,62 +122,64 @@ def breakpoint_add(self, bank, addr):
+            logger.debug(f"Breakpoint remove: {bank:02x}:{addr:04x} {opcode:02x}")
+
+            # Restore opcode
+            if addr < 0x100 and bank == -1:
+                self.bootrom.bootrom[addr] = opcode
+            elif addr < 0x4000:
+                self.cartridge.rombanks[bank, addr] = opcode
+            elif 0x4000 <= addr < 0x8000:
+                self.cartridge.rombanks[bank, addr - 0x4000] = opcode
+            elif 0x8000 <= addr < 0xA000:
+                if bank == 0:
+                    self.lcd.VRAM0[addr - 0x8000] = opcode
+                else:
+                    self.lcd.VRAM1[addr - 0x8000] = opcode
+            elif 0xA000 <= addr < 0xC000:
+                self.cartridge.rambanks[bank, addr - 0xA000] = opcode
+            elif 0xC000 <= addr <= 0xE000:
+                self.ram.internal_ram0[addr - 0xC000] = opcode
             else:
-                self.lcd.VRAM1[addr - 0x8000] = opcode
-        elif 0xA000 <= addr < 0xC000:
-            self.cartridge.rambanks[bank, addr - 0xA000] = opcode
-        elif 0xC000 <= addr <= 0xE000:
-            self.ram.internal_ram0[addr - 0xC000] = opcode
-        else:
-            logger.error("Unsupported breakpoint address. If this a mistake, reach out to the developers")
-        #     return (None, None, None)
-        # return (bank, addr, opcode)
+                logger.error("Unsupported breakpoint address. If this a mistake, reach out to the developers")
+            #     return (None, None, None)
+            # return (bank, addr, opcode)
 
     def breakpoint_reached(self):
-        for i, (bank, pc, _) in enumerate(self.breakpoints_list):
-            if self.cpu.PC == pc and (
-                (pc < 0x4000 and bank == 0 and not self.bootrom_enabled) or \
-                (0x4000 <= pc < 0x8000 and self.cartridge.rombank_selected == bank) or \
-                (0xA000 <= pc < 0xC000 and self.cartridge.rambank_selected == bank) or \
-                (0xC000 <= pc <= 0xFFFF and bank == -1) or \
-                (pc < 0x100 and bank == -1 and self.bootrom_enabled)
-            ):
-                # Breakpoint hit
-                # bank, addr, opcode = self.breakpoint_remove(i)
-                bank, addr, opcode = self.breakpoints_list[i]
-                logger.debug(f"Breakpoint reached: {bank:02x}:{addr:04x} {opcode:02x}")
-                self.breakpoint_waiting = (bank & 0xFF) << 24 | (addr & 0xFFFF) << 8 | (opcode & 0xFF)
-                logger.debug(f"Breakpoint waiting: {self.breakpoint_waiting:08x}")
-                return i
+        pc = self.cpu.PC
+        bank = None
+        if pc < 0x4000 and not self.bootrom_enabled:
+            bank = 0
+        elif 0x4000 <= pc < 0x8000:
+            bank = self.cartridge.rombank_selected
+        elif 0xA000 <= pc < 0xC000:
+            bank = self.cartridge.rambank_selected
+        elif 0xC000 <= pc <= 0xFFFF or (pc <0x100 and self.bootrom_enabled):
+            bank = -1
```

Comment:
There's a potential bug here. But it's my mistake. In other parts of the code, I've used bank 0 for `0xC000 <= pc <= 0xFFFF`.

I'll take a look at this.